accepted/00000-http-api.md

```diff
@@ -0,0 +1,220 @@
+
+Spark provides 3 common ways to attach data to a request. Named route parameters that are part of the base URL, querystring parameters, or the payload itself.
+
+For primitive types, either route parameters or querystring parameters may be used. The desired method shall be specified using a parameter annotation in the handler method signature (explained below).
+
+For complex types, the request payload of type `JSON` must be used. The Spark route wrapper shall unpack the payload using the `GSON` library.
+
+The exposed HTTP routes shall follow the same namespace-method structure with a few extensions:
+
+```
+/rhn/manager/api/namespace/method[/:route/:params][?query=params]
+```
+
+On each request, the Spark route wrapper shall unpack both types of arguments in the request, deserialize the payload, and call the appropriate handler method with unpacked arguments.
+
+To provide a uniform approach to handler registration, Java annotations can be used to specify a method as an endpoint handler. These annotations can be interpreted during initialization to register all the handlers to both frameworks. As a side benefit, all the handler classes in a predetermined package tree can be automatically identified, without the need to add them in a list manually.
+
+Java annotations shall also be used to determine attributes of an endpoint, such as whether it is a read-only method or not.
+
+Below is an example of such an annotation structure:
+
+#### @ApiHandler("namespace") Target=Class
+
+A class marked with this annotation will be registered as the handler of the specified namespace in the XMLRPC framework. The class will further be inspected on initialization to discover any annotated handler methods. Found methods will be registered to Spark as route handlers.
+
+#### @XmlRpcApiEndpoint Target=Method
+#### @HttpApiEndpoint Target=Method
+
+Each of these annotations shall mark a method to act as a handler for one of the APIs.
+
+**HTTP:** These methods will be registered to Spark as route handlers during initialization.
+
+**XMLRPC:** `BaseHandler` will search for a requested method in a list of methods annotated with `@XmlRpcApiEndpoint` in the class during a request.
+
+In most cases, both annotations shall be used together. Exclusive use of any of them shall be reserved for edge cases where the endpoint is not applicable to the specific protocol.
+
+#### @ReadOnly Target=Method
+
+An endpoint handled by this method will be set as a read-only endpoint.
+
+**HTTP:** Spark will create `GET` endpoints for the methods marked with this annotation. If not used, Spark's method wrapper will check the requesting user and block access if restricted.
+
+**XMLRPC:** `BaseHandler` will check this annotation against the requesting user and block access if restricted.
+
+#### @Unauthorized Target=Method
```

Comment:
The role check is already implemented inside of the business logic in all the existing methods so at the moment we don't have any benefit in changing this.

The only exception is a few methods where we don't enforce authentication (e.g. the `login` endpoint itself).

> If we follow the approach I'm proposing in the previous comment, then this can also be a parameter to the annotation.

I think this is a perfectly valid alternative to having a dedicated annotation. It's just a matter of taste, and it is a trivial distinction, so let's decide on this after we see the whole thing demonstrated.

Reply:
I was thinking more in the long term for the project. Decoupling the business logic from the user role. This would simplify the day we decided to refactor the authorization logic in the project.
However, if is already implemented at the method level, we don't have much benefit now for the amount of work it would bring.

Reply:
I agree. If we decide to go over the authorization logic in the future, that would be the best way to implement it.